File: backend_java_spring_boot/myPizza/src/main/java/backendapp/myPizza/services/_SessionService.java
package backendapp.myPizza.services;

import backendapp.myPizza.Models.entities.Order;
import backendapp.myPizza.Models.entities.TimeInterval;
import backendapp.myPizza.Models.entities.WorkSession;
import backendapp.myPizza.Models.enums.OrderStatus;
import backendapp.myPizza.Models.reqDTO.StartSessionDTO;
import backendapp.myPizza.Models.resDTO.ConfirmRes;
import backendapp.myPizza.exceptions.BadRequestException;
import backendapp.myPizza.repositories.OrderRepository;
import backendapp.myPizza.repositories.TimeIntervalRepository;
import backendapp.myPizza.repositories.WorkSessionRepository;
import jakarta.mail.Session;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Service
public class _SessionService {
    @Autowired
    TimeIntervalRepository timeIntervalRp;

    @Autowired
    private OrderRepository orderRp;

    @Autowired
    private WorkSessionRepository _sessionRp;



    public List<Long> getActiveSessionDeliveryTimes() {
        List<Long> list = new ArrayList<>(_sessionRp.getActiveSessionTimeIntervals().stream().map(TimeInterval::getStartsAt).toList());
        list.add(_sessionRp.getActiveSessionTimeIntervals().stream().map(TimeInterval::getEndsAt).toList().getLast());
        return list;
    }

    public List<TimeInterval> getActiveSessionTimeIntervals() {
        return _sessionRp.getActiveSessionTimeIntervals();
    }

    public boolean isThereAnActiveSession() {
        return _sessionRp.findAll().stream().anyMatch(WorkSession::isActive);
    }

    public WorkSession getActiveSession() throws BadRequestException {
        return _sessionRp.findAll().stream().filter(WorkSession::isActive).findFirst().orElseThrow(
                () -> new BadRequestException("There isn't any active session")
        );
    }

    public WorkSession startNewSession(StartSessionDTO startSessionDTO) throws BadRequestException {
        _sessionRp.saveAll(_sessionRp.findAll().stream().peek(s -> {
            if (s.isActive()) s.setActive(false);
        }).toList());
        WorkSession session = new WorkSession(startSessionDTO.type(), startSessionDTO.openTime(), startSessionDTO.closeTime(),
                startSessionDTO.cookCount(), startSessionDTO.ridersCount());
        session.setActive(true);
        _sessionRp.save(session);
        List<TimeInterval> timeIntervals = TimeInterval.getTimeIntervals(startSessionDTO.openTime(), startSessionDTO.closeTime());
        timeIntervalRp.saveAll(timeIntervals);
        for (TimeInterval ti : timeIntervals) {
            session.getTimeIntervals().add(ti);
        }
        return _sessionRp.save(session);
    }

    public ConfirmRes closeActiveSession() throws BadRequestException {
        WorkSession session = getActiveSession();
        session.setActive(false);
        _sessionRp.save(session);
        return new ConfirmRes("Working session with id='" + session.getId() + "' successfully closed", HttpStatus.OK);
    }

    public ConfirmRes addOrderToActiveSession(UUID orderId, UUID timeIntervalId) throws BadRequestException {
        Order order = orderRp.findById(orderId).orElseThrow(
                () -> new BadRequestException("Order wid id =" + orderId + "' doesn't exist")
        );
        WorkSession session = getActiveSession();
        TimeInterval timeInterval = session.getTimeIntervals().stream().filter(ti -> ti.getId().equals(timeIntervalId))
                        .findFirst().orElseThrow(
                        () -> new BadRequestException("Time interval with id='" + timeIntervalId + "' doesn't " +
                                "exist in order with id='" + orderId + "'")
                );
        timeInterval.getOrders().add(order);
        timeIntervalRp.save(timeInterval);
        order.setStatus(OrderStatus.ACCEPTED);
        _sessionRp.save(session);
        return new ConfirmRes("Order wid id ='" + orderId + "' correctly added to active working session", HttpStatus.OK);
    }


}

File: backend_nodejs_nest_fastify/my-pizza/src/nest_modules/order/services/session.service.ts
import { BadRequestException, Injectable, HttpStatus } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { TimeInterval } from '../entities/time-interval.entity';
import { Repository, createQueryBuilder } from 'typeorm';
import { WorkSession } from '../entities/work-session.entity';
import { StartSessionDTO } from '../interfaces/start-session-dto.interface';
import { ConfirmRes } from 'src/nest_modules/auth-user/interfaces/confirm-res.interface';

@Injectable()
export class SessionService {

    constructor(
        @InjectRepository(TimeInterval) private timeIntervalRepository: Repository<TimeInterval>,
        @InjectRepository(WorkSession) private workSessionRepository: Repository<WorkSession>
    ) { }

    public async getPublicSessionDeliveryTimes(): Promise<number[]> {
        const activeSession = await this.getActiveSession()
        if (!activeSession) throw new BadRequestException("There isn't an active session at the moment")
        if (!activeSession.timeIntervals) activeSession.timeIntervals = []
        const deliveryTimes: number[] = []
        activeSession.timeIntervals.forEach(ti => deliveryTimes.push(ti.startsAt))
        deliveryTimes.push(activeSession.timeIntervals[activeSession.timeIntervals.length - 1].endsAt)
        return deliveryTimes
    }

    public async getActiveSessionTimeIntervals(): Promise<TimeInterval[]> {
        return (await this.getActiveSession()).timeIntervals
    }

    public async isThereAnActiveSession(): Promise<boolean> {
        return !!await this.getActiveSession()
    }

    public async getActiveSession(): Promise<WorkSession | null | undefined> {
        const activeSession = await this.workSessionRepository.createQueryBuilder('session')
            .where('session.active = true')
            .getOne()
        const timeIntervals = await this.timeIntervalRepository.createQueryBuilder('ti')
            .where('ti.workSessionId = :id', { id: activeSession.id })
            .getMany()
        timeIntervals.map(ti => {
            if (!ti.orders) ti.orders = []
            return ti
        })
        activeSession.timeIntervals = timeIntervals
        return activeSession
    }

    public async startNewSession(startSessionDTO: StartSessionDTO): Promise<WorkSession> {
        await this.workSessionRepository.save((await this.workSessionRepository.createQueryBuilder('session')
            .where('session.active = true')
            .getMany()
        ).map(session => {
            if (session.active) session.active = false
            return session
        }))
        const session = new WorkSession(startSessionDTO.type, startSessionDTO.openTime,
            startSessionDTO.closeTime, startSessionDTO.cookCount, startSessionDTO.ridersCount)
        session.active = true
        await this.workSessionRepository.save(session)
        const timeIntervals = TimeInterval.getTimeIntervals(startSessionDTO.openTime, startSessionDTO.closeTime)
        await this.timeIntervalRepository.save(timeIntervals)
        session.timeIntervals = timeIntervals
        return await this.workSessionRepository.save(session)
    }

    public async closeActiveSession(): Promise<ConfirmRes> {
        const activeSession = await this.getActiveSession()
        activeSession.active = false
        await this.workSessionRepository.save(activeSession)
        return {
            message: "Working session with id='" + activeSession.id + "' successfully closed",
            timestamp: new Date().getTime(),
            statusCode: HttpStatus.OK
        }
    }

}
